elastic/apm-server
# sampling: fix pubsub implementation

## Motivation/summary

The initial implementation was written as a ~quick hack, with the expectation that it would be replaced by the Changes API.
It was broken due to its ignorance of data streams, and multi-shard indices. Sequence numbers are only comparable within a single shard.

Instead of waiting for the Changes API we could take the same approach as https://github.com/elastic/fleet-server/pull/200, using
a new Elasticsearch API built for Fleet Server. The main issue here is that the API does not (currently) support data streams.

Given the above and that there is no known delivery date for the Changes API, we propose to instead revise the pubsub implementation to address the problems by:

 - enforcing single-shard indices for sampled trace data streams
 - searching (now single-shard) backing indices individually

In addition, we now use global checkpoints to bound searches (ensuring all replicas have committed the documents). Querying underlying indices and global checkpoints will require an additional "monitor" index privilege.

## Checklist

~- [ ] Documentation has been updated~

## How to test these changes

1. Run two APM Servers with tail-based sampling enabled
2. Create some transactions
3. Force a [rollover](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-rollover-index.html) of the sampled traces data stream, to create a new index
4. Create some more transactions
5. Check transactions are indexed according to the sampling config

## Related issues

Closes https://github.com/elastic/apm-server/issues/5119
